service/history/workflow/context.go

```diff
@@ -143,6 +144,15 @@ type (
 )
 
 type (
+	pinnedState struct {
+		updateRegistry update.Registry
+	}
+
+	pinKey struct {
+		wfkey  definition.WorkflowKey
+		pinner string
```

Comment:
Can't just assume there will only be a single instance of a type - context.Shard will have pinned update.Registry instances from hundreds/thousands of workflow.Contexts. Can't assume that definition.WorkflowKey is sufficient because some other part of workflow processing may _also_ need to pin some memory. So I end up using the workflow identifier (definition.WorkflowKey) plus the name of the type doing the pinning ("workflow.ContextImpl") to get full uniqueness in the set of keys.